src/objective/rank_objective.hpp

```diff
@@ -178,43 +232,87 @@ class LambdarankNDCG : public RankingObjective {
           high_rank = j;
           low_rank = i;
         }
-        const data_size_t high = sorted_idx[high_rank];
-        const int high_label = static_cast<int>(label[high]);
-        const double high_score = score[high];
-        const double high_label_gain = label_gain_[high_label];
-        const double high_discount = DCGCalculator::GetDiscount(high_rank);
+
+        // info of more relevant doc
+        const data_size_t high = sorted_idx[high_rank];  // doc index in query results
+        const int high_label = static_cast<int>(label[high]);  // label (Y)
+        const double high_score = score[high];  // current model predicted score
+        const double high_label_gain = label_gain_[high_label];  // default: 2^high_label - 1
+        const double high_discount = DCGCalculator::GetDiscount(high_rank);  // 1/log2(2 + i)
+
+        // info of less relevant doc
         const data_size_t low = sorted_idx[low_rank];
         const int low_label = static_cast<int>(label[low]);
         const double low_score = score[low];
         const double low_label_gain = label_gain_[low_label];
         const double low_discount = DCGCalculator::GetDiscount(low_rank);
 
+        //
+        // note on subsequent comments
+        // in the papers, customary to assume i is more relevant than j
+        // formula numbers are from unbiased lambdamart paper
+        //
+        // si - sj
         const double delta_score = high_score - low_score;
 
         // get dcg gap
+        // default: 2^i - 2^j > 0
         const double dcg_gap = high_label_gain - low_label_gain;
+
         // get discount of this pair
+        // |1/log2(2 + i) - 1/log2(2 + j)|
         const double paired_discount = fabs(high_discount - low_discount);
+
         // get delta NDCG
+        // (2^i - 2^j) * |1/log2(2 + i) - 1/log2(2 + j)| / max_dcg
         double delta_pair_NDCG = dcg_gap * paired_discount * inverse_max_dcg;
-        // regular the delta_pair_NDCG by score distance
-        if (norm_ && best_score != worst_score) {
+
+        // regularize the delta_pair_NDCG by score distance
+        if ((norm_ || unbiased_) && best_score != worst_score) {
           delta_pair_NDCG /= (0.01f + fabs(delta_score));
         }
+
         // calculate lambda for this pair
-        double p_lambda = GetSigmoid(delta_score);
+        // part of (34)
+        // (34) and (36) are used to get the unbiased gradient estimates
+        double p_lambda = GetSigmoid(delta_score);  // 1 / (1 + e^(sigmoid_ * (si - sj)))
+
+        // d/dx {part of (34)} from above
         double p_hessian = p_lambda * (1.0f - p_lambda);
```

Comment:
q: this line is unchanged, but it's diff in [this](https://github.com/acbull/Unbiased_LambdaMart/blob/master/Unbias_LightGBM/src/objective/rank_objective.hpp#L197) line in the original repo. what's up?
a: again, `sigmoid_ = 2` stuff.